presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java

```diff
@@ -13,34 +13,81 @@
 import com.facebook.presto.Session;
 import com.facebook.presto.common.plan.PlanCanonicalizationStrategy;
+import com.facebook.presto.spi.plan.AggregationNode;
 import com.facebook.presto.spi.plan.PlanNode;
 import com.facebook.presto.spi.plan.PlanNodeWithHash;
 import com.facebook.presto.spi.statistics.HistoricalPlanStatistics;
 import com.facebook.presto.spi.statistics.HistoryBasedPlanStatisticsProvider;
 import com.facebook.presto.spi.statistics.HistoryBasedSourceInfo;
+import com.facebook.presto.spi.statistics.PlanStatistics;
 import com.facebook.presto.sql.planner.PlanHasher;
 import com.facebook.presto.sql.planner.TypeProvider;
 import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.Weigher;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import io.airlift.units.DataSize;
 
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Supplier;
 
 import static com.facebook.presto.SystemSessionProperties.useHistoryBasedPlanStatisticsEnabled;
 import static com.facebook.presto.common.plan.PlanCanonicalizationStrategy.historyBasedPlanCanonicalizationStrategyList;
 import static com.facebook.presto.sql.planner.iterative.Plans.resolveGroupReferences;
 import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.graph.Traverser.forTree;
 import static java.util.Objects.requireNonNull;
 
 public class HistoryBasedPlanStatisticsCalculator
         implements StatsCalculator
 {
+    private static final Logger LOG = Logger.get(HistoryBasedPlanStatisticsCalculator.class);
+
+    private static final List<Class<? extends PlanNode>> PRECOMPUTE_PLAN_NODES = ImmutableList.of(JoinNode.class, SemiJoinNode.class, AggregationNode.class);
+    private static final DataSize CACHE_SIZE_BYTES = new DataSize(2, DataSize.Unit.MEGABYTE);
+
+    // For weight, we only consider size of hash, as PlanNodes are already in memory for running queries.
+    // We use length of hash + 20 bytes to account for stats.
+    private final LoadingCache<PlanNodeWithHash, HistoricalPlanStatistics> cache = CacheBuilder.newBuilder()
+            .maximumWeight(CACHE_SIZE_BYTES.toBytes())
+            .weigher((Weigher<PlanNodeWithHash, HistoricalPlanStatistics>) (key, statistics) -> key.getHash().orElse("").length() + 20)
+            .expireAfterWrite(5, TimeUnit.MINUTES)
```

Comment:
Would it make sense to make the 5 minutes configurable?

Reply:
Its ok for now, I plan to remove this in future PRs and cache only when query is running, and then drop